File: packages/js/cli/src/browse.js
/**
 * CLI helpers for `handoffkit-js browse …`
 */
export async function runBrowseCommand(argv = [], io = {}) {
  const stdout = io.stdout || ((text) => console.log(text));
  const {
    createBrowserAgentKit,
    makeFixtureMapTransport,
    webSearch,
    gatherWebResearch,
    researchPromptSection,
  } = await import("@handoffkit/browser");

  const [sub, ...rest] = argv;
  if (!sub || sub === "--help" || sub === "-h") {
    stdout([
      "handoffkit-js browse",
      "",
      "  browse search <query> [--max 6] [--json] [--allow host] [--deny host]",
      "  browse fetch <url> [--markdown|--json] [--format markdown|readme]",
      "  browse research <query> [--max-pages 3] [--json|--markdown] [--cache] [--allow host]",
      "  browse deep <query> [--max-pages 8] [--max-depth 2] [--json|--markdown] [--cache]",
      "  browse fixture",
      "  browse tools",
    ].join("\n"));
    return 0;
  }

  const asJson = rest.includes("--json");
  const asMarkdown = rest.includes("--markdown");
  const useCache = rest.includes("--cache");
  const useFixture = rest.includes("--fixture") || rest.includes("--transport=fixture");
  const max = Number(readBrowseFlag(rest, "--max") || readBrowseFlag(rest, "--max-pages") || 0) || undefined;
  const maxDepth = Number(readBrowseFlag(rest, "--max-depth") || 0) || undefined;
  const format = readBrowseFlag(rest, "--format") || "markdown";
  const allowHosts = collectBrowseFlags(rest, "--allow");
  const denyHosts = collectBrowseFlags(rest, "--deny");

  const kit = createBrowserAgentKit({
    fixture: useFixture,
    useCache,
    allowHosts,
    denyHosts,
    format,
    maxPages: max || 3,
  });

  if (sub === "tools") {
    stdout(kit.tools.map((t) => t.name).join("\n"));
    return 0;
  }

  if (sub === "fixture") {
    const transport = makeFixtureMapTransport();
    const pack = await gatherWebResearch({
      seedUrls: ["https://fixture.local/"],
      seedOnly: true,
      transport,
      maxPages: 3,
      preferExplore: true,
      maxDepth: 1,
    });
    stdout(asJson ? JSON.stringify(pack.toDict(), null, 2) : pack.markdown_context || researchPromptSection(pack));
    return pack.pages_ok > 0 ? 0 : 1;
  }

  if (sub === "search") {
    const query = rest.filter((a) => !a.startsWith("--")).join(" ").trim();
    if (!query) throw new Error("browse search requires a query.");
    const result = await webSearch(query, {
      transport: kit.transport,
      maxResults: max || 6,
      allowHosts,
      denyHosts,
    });
    if (asJson) stdout(JSON.stringify(result, null, 2));
    else {
      stdout(`query: ${result.query}`);
      stdout(`count: ${result.count}`);
      for (const hit of result.results) {
        stdout(`- ${hit.title || "(untitled)"} :: ${hit.url}`);
      }
    }
    return result.success ? 0 : 1;
  }

  if (sub === "fetch") {
    const url = rest.find((a) => !a.startsWith("--"));
    if (!url) throw new Error("browse fetch requires a url.");
    const page = await kit.fetchMarkdown(url, { format, maxChars: 60000 });
    if (asJson) stdout(JSON.stringify(page.toDict(), null, 2));
    else stdout(page.markdown || page.error || "");
    return page.success ? 0 : 1;
  }

  if (sub === "research") {
    const query = rest.filter((a) => !a.startsWith("--")).join(" ").trim();
    if (!query) throw new Error("browse research requires a query.");
    const pack = await kit.gather({
      query,
      maxPages: max || 3,
      allowHosts,
      denyHosts,
      format,
      useCache,
    });
    if (asJson) stdout(JSON.stringify(pack.toDict(), null, 2));
    else if (asMarkdown) stdout(pack.markdown_context);
    else stdout(researchPromptSection(pack) || pack.error || "no research");
    return pack.pages_ok > 0 ? 0 : 1;
  }

  if (sub === "deep") {
    const query = rest.filter((a) => !a.startsWith("--")).join(" ").trim();
    if (!query) throw new Error("browse deep requires a query.");
    const pack = await kit.deepGather({
      query,
      maxPages: max || 8,
      maxDepth: maxDepth ?? 2,
      seedUrls: useFixture ? ["https://fixture.local/"] : [],
      autoSearch: !useFixture,
      allowHosts,
      denyHosts,
      format,
      useCache,
    });
    if (asJson) stdout(JSON.stringify(pack.toDict(), null, 2));
    else if (asMarkdown) stdout(pack.markdown_context);
    else stdout(researchPromptSection(pack) || pack.error || "no deep research");
    return pack.pages_ok > 0 ? 0 : 1;
  }

  throw new Error(`Unknown browse subcommand: ${sub}`);
}

function readBrowseFlag(argv, name) {
  const index = argv.indexOf(name);
  if (index === -1) return null;
  const value = argv[index + 1];
  if (!value || value.startsWith("--")) throw new Error(`${name} requires a value.`);
  return value;
}

function collectBrowseFlags(argv, name) {
  const out = [];
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === name) {
      const value = argv[i + 1];
      if (!value || value.startsWith("--")) throw new Error(`${name} requires a value.`);
      out.push(value);
      i += 1;
    }
  }
  return out;
}

File: python/packages/handoffkit/handoffkit/_cli/browse.py
"""CLI helpers for ``handoffkit browse …``."""

from __future__ import annotations

import argparse
import json
from typing import Any


def add_browse_parser(subparsers: Any) -> None:
    browse = subparsers.add_parser(
        "browse",
        help="First-party web search / fetch / research (handoffkit.browser).",
    )
    browse_sub = browse.add_subparsers(dest="browse_command")

    search_p = browse_sub.add_parser("search", help="Live/offline web search.")
    search_p.add_argument("query", nargs="+", help="Search query.")
    search_p.add_argument("--max", type=int, default=6)
    search_p.add_argument("--json", action="store_true")
    search_p.add_argument("--fixture", action="store_true")
    search_p.add_argument("--allow", action="append", default=[])
    search_p.add_argument("--deny", action="append", default=[])

    fetch_p = browse_sub.add_parser("fetch", help="Fetch one URL as markdown.")
    fetch_p.add_argument("url", help="Absolute http(s) URL.")
    fetch_p.add_argument("--json", action="store_true")
    fetch_p.add_argument("--markdown", action="store_true")
    fetch_p.add_argument("--format", choices=["markdown", "readme"], default="markdown")
    fetch_p.add_argument("--fixture", action="store_true")

    research_p = browse_sub.add_parser("research", help="Search-then-fetch research pack.")
    research_p.add_argument("query", nargs="+", help="Research query.")
    research_p.add_argument("--max-pages", type=int, default=3)
    research_p.add_argument("--json", action="store_true")
    research_p.add_argument("--markdown", action="store_true")
    research_p.add_argument("--cache", action="store_true")
    research_p.add_argument("--fixture", action="store_true")
    research_p.add_argument("--allow", action="append", default=[])
    research_p.add_argument("--deny", action="append", default=[])
    research_p.add_argument("--format", choices=["markdown", "readme"], default="markdown")

    fixture_p = browse_sub.add_parser("fixture", help="Offline fixture crawl demo.")
    fixture_p.add_argument("--json", action="store_true")
    browse_sub.add_parser("tools", help="List registered browser tool names.")


def run_browse_command(args: argparse.Namespace) -> int:
    from handoffkit.browser import (
        create_browser_agent_kit,
        gather_web_research,
        make_fixture_map_transport,
        research_prompt_section,
        web_search,
    )

    cmd = getattr(args, "browse_command", None)
    if not cmd:
        print(
            "handoffkit browse\n\n"
            "  browse search <query> [--max 6] [--json] [--allow host] [--deny host]\n"
            "  browse fetch <url> [--markdown|--json] [--format markdown|readme]\n"
            "  browse research <query> [--max-pages 3] [--json|--markdown] [--cache]\n"
            "  browse fixture\n"
            "  browse tools"
        )
        return 0

    use_fixture = bool(getattr(args, "fixture", False))
    kit = create_browser_agent_kit(
        {
            "fixture": use_fixture,
            "use_cache": bool(getattr(args, "cache", False)),
            "allow_hosts": list(getattr(args, "allow", []) or []),
            "deny_hosts": list(getattr(args, "deny", []) or []),
            "format": getattr(args, "format", "markdown"),
            "max_pages": int(getattr(args, "max_pages", 3) or 3),
        }
    )

    if cmd == "tools":
        print("\n".join(t.name for t in kit["tools"]))
        return 0

    if cmd == "fixture":
        transport = make_fixture_map_transport()
        pack = gather_web_research(
            transport=transport,
            seed_urls=["https://fixture.local/"],
            seed_only=True,
            max_pages=3,
            prefer_explore=True,
            max_depth=1,
        )
        if getattr(args, "json", False):
            print(json.dumps(pack.to_dict(), indent=2, ensure_ascii=False))
        else:
            print(pack.markdown_context or research_prompt_section(pack))
        return 0 if pack.pages_ok > 0 else 1

    if cmd == "search":
        query = " ".join(args.query).strip()
        result = web_search(
            query,
            transport=kit["transport"],
            max_results=int(getattr(args, "max", 6) or 6),
            allow_hosts=list(getattr(args, "allow", []) or []),
            deny_hosts=list(getattr(args, "deny", []) or []),
        )
        if args.json:
            print(json.dumps(result, indent=2, ensure_ascii=False))
        else:
            print(f"query: {result['query']}")
            print(f"count: {result['count']}")
            for hit in result.get("results") or []:
                print(f"- {hit.get('title') or '(untitled)'} :: {hit.get('url')}")
        return 0 if result.get("success") else 1

    if cmd == "fetch":
        page = kit["fetch_markdown"](args.url, format=getattr(args, "format", "markdown"))
        if args.json:
            print(json.dumps(page.to_dict(), indent=2, ensure_ascii=False))
        else:
            print(page.markdown or page.error or "")
        return 0 if page.success else 1

    if cmd == "research":
        query = " ".join(args.query).strip()
        pack = kit["gather"](
            query=query,
            max_pages=int(getattr(args, "max_pages", 3) or 3),
            allow_hosts=list(getattr(args, "allow", []) or []),
            deny_hosts=list(getattr(args, "deny", []) or []),
            format=getattr(args, "format", "markdown"),
        )
        if args.json:
            print(json.dumps(pack.to_dict(), indent=2, ensure_ascii=False))
        elif getattr(args, "markdown", False):
            print(pack.markdown_context or "")
        else:
            print(research_prompt_section(pack) or pack.markdown_context or pack.error)
        return 0 if pack.pages_ok > 0 else 1

    print(f"unknown browse subcommand: {cmd}")
    return 1
